src/table-element.ts

```diff
@@ -28,18 +27,33 @@ export class TableElement<T> {
   }
 
   startEdit(): void {
-    this.originalData = cloneDeep(this.currentData);
-    this.editing = true;
+    this.editingState = { kind: StateKind.Editing, originalData: cloneDeep(this.currentData) }
     this.validator.enable();
   }
 
   cancelOrDelete(): void {
-    if (this.id == -1 || !this.editing)
+    if (this.id == -1 || this.editingState.kind === StateKind.NotEditing)
       this.delete();
     else {
-      this.currentData = this.originalData;
-      this.editing = false;
+      this.currentData = this.editingState.originalData;
```

Comment:
I've been thinking about this approach, and I personally feel that we are "forcing the code" to make it work. What we need to ensure here is to have any value when doing `this.currentData = this.originalData;`. If there is no data on `this.originalData`, then we should raise an error, because it's value shouldn't be null (that is what we want to ensure).

What do you think on doing something like:
```if (!this.originalData)
        throw new Error('There is no original data to swap.');```

Reply:
With the code we have in the PR, we already have a compile-time check that it's _impossible_ that such case will happen. For example, if we mistakenly write this code:

```
if (this.id == -1) // i.e., forgetting the condition of not-editing state
  this.delete();
else {
  // that way it's possible that we're in editing or non-editing state and so originalData may be non-existent
  this.currentData = this.editingState.originalData;
 //...
}
```
we'll get a compile-time error, which is very nice. That was my intention; to have something like a proof or a compile-time check that we're never doing `this.currentData = this.originalData;` if there's no data on `this.originalData`.

As for your suggestion: think about the cases when this error will be thrown. These cases are cases in which we _mistakenly_ wrote `this.currentData = this.originalData;` when we shouldn't have. What throwing the error will do is failing early/fast. But it didn't ensure that we didn't make such a wrong assignment in the first place. That's my opinion.

But, if for the sake of keeping backward compatibility, adding the check and throwing the error is, of course, better than failing later. But I don't think it gives the same guarantees the code in the PR gives.

Reply:
No problem at all, @irossimoline , thank _you_ for the awesome package :)

<hr/>

Ummm ... okay. After the edit, I see what you mean. That way we can pass `--strictNullChecks`. But still, the code with the if-condition leaves the possibility that we throw the exception, while the code in the PR has a compile-time check that it'll never happen.

For example, let's say we write this totally-wrong code:

```
this.editing = false;
this.originalData = null;
if (!this.originalData)
  throw new Error('There is no original data to swap.');
this.currentData = this.originalData;
```

The code above will throw in runtime. But the code below will not even compile (which is better):

```
this.editingState = { kind: StateKind.NotEditing };
this.currentData = this.editingState.originalData; // compile-time error!
```

In this code, the compiler inferred that `this.editingState` if of type `NotEditing`, and the type `NotEditing` doesn't have a member called `originalData`.

<hr/>

> We are defining originalData in Editing class as not nullable, but we are leaving it as null/undefined when instanciating it

Nope, we're not. In line-116, we're instantiating a member of `NotEditing`.

An equivalent way of writing the code would be applying these replacements:

`{ kind: StateKind.NotEditing }` -> `new NotEditing(StateKind.NotEditing)`  
`{ kind: StateKind.Editing, originalData: x }` -> `new Editing(StateKind.Editing, x)`

I.e., for example, you could _not_ write `new NotEditing(StateKind.Editing)`. Or in another words: `{ kind: StateKind.NotEditing }` is _inferred_ to be of `NotEditing` class (and similarly for `Editing` class).

(side note: I know sending `StateKind.NotEditing` to `NotEditing` constructor is boilerplate, it's my mistake because of how I wrote the code. I could re-write it so that we could call `new NotEditing()` and `new Editing(x)` directly)

<hr/>

> With the if condition we are ensuring that the code doesn't fail on compile-time.

Nope, we're not ensuring that. As an example, you can check the example above.

> The compile error is gone when using the if condition mentioned previously, because typescript detects that we are no longer assigning T? to T, but T to T.

Indeed, the compiler error is gone, but our code could fail in runtime. Also, explained in the example above.

<hr/>

So, for summary: No, I don't see the 2 ways as equivalent, and for the decision whether or not to break backward compatibility, I'll leave that to you. I agree backward compatibility could be important.

Needless to say, at any moment, if you see as too time-consuming, just go with what you see best. Honestly, I'm discussing it more from the standpoint of "[Discriminated Union Types](http://www.typescriptlang.org/docs/handbook/advanced-types.html) are great! I hope you see their value and know how to use them well" more than "what's the most suitable decision for this particular case?".